Add ctor/dtor/copy/assign to std::vector bindings

omerfirmak opened the pull request.

omerfirmak committed:
Add ctor/dtor/copy/assign to std::vector bindings

omerfirmak commented:
```
submodules/serialization/source/agora/serialization/Serializer.d(698,17): Error: none of the overloads of __ctor are callable using a const object, candidates are:
source/scpd/Cpp.d(62,9):        scpd.types.Stellar_SCP.SCPBallot.CPPBindingMixin!(SCPBallot, true, true).this(CppCtor use)
source/scpd/Cpp.d(57,13):        scpd.types.Stellar_SCP.SCPBallot.CPPBindingMixin!(SCPBallot, true, true).this(uint, xvector!(ubyte, 4294967292u))
source/scpd/Cpp.d(73,9):        scpd.types.Stellar_SCP.SCPBallot.CPPBindingMixin!(SCPBallot, true, true).this(return ref scope inout(SCPBallot) rhs)

```

Shouldnt this work with `const` objects?

https://github.com/bosagora/agora/pull/2081/files#diff-2fff5f741967021f44722f88e37effa07123b287d13e843ff246b43d91a7d8d4R57-R60

omerfirmak commented:
This mostly works apart from a random crash that happens during GC finalizing a `vector`

omerfirmak commented:
> This mostly works apart from a random crash that happens during GC finalizing a `vector`

I can't reproduce this after a rebase but I don't see where we would GC allocate a `SCPEnvelope` anyways. Marking this ready for review for the time being.

Geod24 commented on source/scpd/Cpp.d:
```diff
@@ -37,44 +37,44 @@ import vibe.data.json;
-    void opAssignCPPObject(T) (T* lhs, T* rhs);
-    void copyCtorCPPObject(T) (T* ptr, inout(T)* rhs);
-    void copyCtorCPPObject(T) (immutable(T)* ptr, inout(T)* rhs);
+@trusted nothrow @nogc pure:
+    void defaultCtorCPPObject(T) (inout(T)* ptr);
+    void dtorCPPObject(T) (inout(T)* ptr);
+    void opAssignCPPObject(T) (inout(T)* lhs, inout(T)* rhs);
+    void copyCtorCPPObject(T) (inout(T)* ptr, inout(T)* rhs);
```
Not sure why you would need `inout` for single-argument templated free functions, but I'm sure there's an horror story behind this :)

Geod24 commented on source/scpd/Cpp.d:
```diff
@@ -37,44 +37,44 @@ import vibe.data.json;
+        import std.traits : Fields, FieldNameTuple;
+
+        static if (!DefaultConstructable) 
+            @disable this();
```
```suggestion
            @disable this ();
```
👼

Geod24 commented on source/scpd/Cpp.d:
```diff
@@ -37,44 +37,44 @@ import vibe.data.json;
         {
+            // Without this, compiler thinks we are not initializing anything
+            static foreach (field; FieldNameTuple!T)
+                __traits(getMember, this, field) = __traits(getMember, this, field).init;
```
Would it be possible to nest it in a dead condition just so that we don't run code that's not needed ?

omerfirmak replied:
Yeah that worked.

omerfirmak committed:
Ensure no invalid SCPStatement can be introduced in to the system

through deserialization.

This makes sure that even if we receive a bad statement from an API
we can gracefully handle that. Other sources that may introduce a bad
statement should still probably crash the node as they are most likely
programming errors.

Fixes #657

omerfirmak commented:
Rebased

Geod24 reviewed (approved):
Noice

Geod24 commented:
Reverted in #2143
